File: editor-server/src/resolvers/subscriptions/dancer.ts
import { ObjectType, Field, registerEnumType } from "type-graphql";
import { Dancer } from "../../../prisma/generated/type-graphql";

@ObjectType()
export class DancerPayload {
  @Field((type) => dancerMutation)
    mutation: dancerMutation;

  @Field()
    editBy: number;

  @Field((type) => Dancer, { nullable: true })
    dancerData?: Dancer | null;
}

export enum dancerMutation {
  UPDATED = "UPDATED",
  CREATED = "CREATED",
  DELETED = "DELETED",
}

registerEnumType(dancerMutation, {
  name: "dancerMutation",
});

File: editor-server/src/graphql/subscriptions/dancer.rs
//! Dancer subscription methods.

use crate::graphql::subscriptor::Subscriptor;
use crate::graphql::types::dancer::Dancer;

use async_graphql::{Enum, SimpleObject, Subscription};
use futures_core::stream::Stream;
use serde::{Deserialize, Serialize};

#[derive(Enum, Clone, Copy, Eq, PartialEq, Default, Serialize, Deserialize)]
pub enum DancerMutationMode {
    #[default]
    #[serde(rename = "UPDATED")]
    Updated,
    #[serde(rename = "CREATED")]
    Created,
    #[serde(rename = "DELETED")]
    Deleted,
}

#[derive(SimpleObject, Clone, Default)]
pub struct DancerPayload {
    pub mutation: DancerMutationMode,
    pub dancer_data: Option<Dancer>,
    pub edit_by: i32,
}

#[derive(Default)]
pub struct DancerSubscription;

#[Subscription]
impl DancerSubscription {
    async fn dancer_subscription(&self) -> impl Stream<Item = DancerPayload> {
        Subscriptor::<DancerPayload>::subscribe()
    }
}
